File: specs/001-architecture-core-ddd/contracts/core-types-contract.py
"""
Architecture.Core - Python Django Implementation
API Contract Specification

This file defines the complete contract for DDD abstractions and functional types
that must be implemented for Python Django. These contracts ensure consistency
across all language implementations while leveraging Python-specific features.

Date: 2025-09-22
Language: Python 3.12+
Framework: Django 4+ (optional integration)
"""

from typing import TypeVar, Generic, Protocol, Callable, Iterator, Optional, List, Dict, Any, Union
from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import Enum
from datetime import datetime
import asyncio

# ============================================================================
# ENTITY IDENTIFICATION CONTRACTS
# ============================================================================

@Protocol
class EntityId:
    """Protocol defining contract for entity identifiers"""

    def __str__(self) -> str:
        """String representation of the ID"""
        ...

    def __eq__(self, other: object) -> bool:
        """Equality comparison with other objects"""
        ...

    def __hash__(self) -> int:
        """Hash code for use in collections"""
        ...

# ============================================================================
# DOMAIN LAYER CONTRACTS
# ============================================================================

TId = TypeVar('TId', bound=EntityId)
TAggregate = TypeVar('TAggregate', bound='AggregateRoot')

class DomainEvent(Protocol):
    """Contract for domain events"""

    @property
    def id(self) -> str:
        """Unique identifier for the event"""
        ...

    @property
    def occurred_at(self) -> datetime:
        """When the event occurred"""
        ...

    @property
    def correlation_id(self) -> Optional[str]:
        """Correlation ID for event tracking"""
        ...

    @property
    def causation_id(self) -> Optional[str]:
        """Causation ID for event tracking"""
        ...

@dataclass(frozen=True)
class DomainEventBase:
    """Base implementation for domain events"""
    id: str
    occurred_at: datetime
    correlation_id: Optional[str] = None
    causation_id: Optional[str] = None

class AggregateRoot(Generic[TId], ABC):
    """Contract for aggregate roots"""

    def __init__(self, id: TId, version: int = 0) -> None:
        """Initialize aggregate with ID and version"""
        ...

    @property
    def id(self) -> TId:
        """Immutable aggregate identifier"""
        ...

    @property
    def version(self) -> int:
        """Version for optimistic concurrency control"""
        ...

    @property
    def domain_events(self) -> List[DomainEvent]:
        """Read-only collection of domain events"""
        ...

    def add_event(self, event: DomainEvent) -> None:
        """Add domain event to the aggregate"""
        ...

    def clear_events(self) -> None:
        """Clear all domain events"""
        ...

    def increment_version(self) -> None:
        """Increment version for optimistic concurrency"""
        ...

class Entity(Generic[TId], ABC):
    """Contract for domain entities"""

    def __init__(self, id: TId) -> None:
        """Initialize entity with ID"""
        ...

    @property
    def id(self) -> TId:
        """Immutable entity identifier"""
        ...

    def __eq__(self, other: object) -> bool:
        """Identity-based equality comparison"""
        ...

    def __hash__(self) -> int:
        """Hash based on entity ID"""
        ...

class ValueObject(ABC):
    """Contract for value objects"""

    @abstractmethod
    def get_equality_components(self) -> Iterator[Any]:
        """Return components used for equality comparison"""
        ...

    def __eq__(self, other: object) -> bool:
        """Structural equality based on components"""
        ...

    def __hash__(self) -> int:
        """Hash based on equality components"""
        ...

# ============================================================================
# FUNCTIONAL TYPES CONTRACTS
# ============================================================================

T = TypeVar('T')
U = TypeVar('U')

class ErrorCategory(Enum):
    """Categories for error classification"""
    DOMAIN = "Domain"
    VALIDATION = "Validation"
    INFRASTRUCTURE = "Infrastructure"
    CONCURRENCY = "Concurrency"
    SECURITY = "Security"

@dataclass(frozen=True)
class Error:
    """Contract for structured error information"""
    code: str
    message: str
    category: ErrorCategory
    metadata: Dict[str, Any]

    @classmethod
    def domain(cls, code: str, message: str) -> 'Error':
        """Create domain error"""
        ...

    @classmethod
    def validation(cls, code: str, message: str, metadata: Dict[str, Any] = None) -> 'Error':
        """Create validation error"""
        ...

    @classmethod
    def infrastructure(cls, code: str, message: str, cause: Exception = None) -> 'Error':
        """Create infrastructure error"""
        ...

    @classmethod
    def concurrency(cls, code: str, message: str) -> 'Error':
        """Create concurrency error"""
        ...

    @classmethod
    def security(cls, code: str, message: str) -> 'Error':
        """Create security error"""
        ...

class Result(Generic[T]):
    """Contract for functional result type"""

    @classmethod
    def success(cls, value: T) -> 'Result[T]':
        """Create successful result"""
        ...

    @classmethod
    def failure(cls, error: Error) -> 'Result[T]':
        """Create failed result"""
        ...

    @property
    def is_success(self) -> bool:
        """Check if result is successful"""
        ...

    @property
    def is_failure(self) -> bool:
        """Check if result is failed"""
        ...

    @property
    def value(self) -> T:
        """Get success value (raises if failed)"""
        ...

    @property
    def error(self) -> Error:
        """Get error (raises if successful)"""
        ...

    def map(self, func: Callable[[T], U]) -> 'Result[U]':
        """Map function over success value"""
        ...

    def bind(self, func: Callable[[T], 'Result[U]']) -> 'Result[U]':
        """Monadic bind operation"""
        ...

    def match(self, on_success: Callable[[T], U], on_failure: Callable[[Error], U]) -> U:
        """Pattern matching for result"""
        ...

    def ensure(self, predicate: Callable[[T], bool], error: Error) -> 'Result[T]':
        """Ensure predicate holds for success value"""
        ...

    @staticmethod
    def combine(results: List['Result[T]']) -> 'Result[List[T]]':
        """Combine multiple results into single result"""
        ...

class Maybe(Generic[T]):
    """Contract for optional value type"""

    @classmethod
    def some(cls, value: T) -> 'Maybe[T]':
        """Create Some with value"""
        ...

    @classmethod
    def none(cls) -> 'Maybe[T]':
        """Create None value"""
        ...

    @property
    def has_value(self) -> bool:
        """Check if value is present"""
        ...

    @property
    def value(self) -> T:
        """Get value (raises if None)"""
        ...

    def map(self, func: Callable[[T], U]) -> 'Maybe[U]':
        """Map function over Some value"""
        ...

    def bind(self, func: Callable[[T], 'Maybe[U]']) -> 'Maybe[U]':
        """Monadic bind operation"""
        ...

    def or_else(self, default: T) -> T:
        """Get value or default"""
        ...

    def or_else_get(self, func: Callable[[], T]) -> T:
        """Get value or call function for default"""
        ...

    def filter(self, predicate: Callable[[T], bool]) -> 'Maybe[T]':
        """Filter value by predicate"""
        ...

    @staticmethod
    def from_optional(value: Optional[T]) -> 'Maybe[T]':
        """Create Maybe from Optional value"""
        ...

# ============================================================================
# REPOSITORY CONTRACTS
# ============================================================================

class Repository(Generic[TAggregate, TId], Protocol):
    """Contract for aggregate repositories"""

    async def get_by_id_async(self, id: TId) -> Maybe[TAggregate]:
        """Get aggregate by ID asynchronously"""
        ...

    async def add_async(self, aggregate: TAggregate) -> Result[None]:
        """Add new aggregate asynchronously"""
        ...

    async def update_async(self, aggregate: TAggregate) -> Result[None]:
        """Update existing aggregate asynchronously"""
        ...

    async def delete_async(self, id: TId) -> Result[None]:
        """Delete aggregate by ID asynchronously"""
        ...

    async def exists_async(self, id: TId) -> bool:
        """Check if aggregate exists asynchronously"""
        ...

# ============================================================================
# DJANGO INTEGRATION CONTRACTS
# ============================================================================

class DjangoModelMixin:
    """Contract for Django model integration"""

    version: int
    created_at: datetime
    updated_at: datetime

    def increment_version(self) -> None:
        """Increment version for optimistic concurrency"""
        ...

class DjangoRepository(Repository[TAggregate, TId], ABC):
    """Contract for Django-specific repository implementations"""

    def __init__(self, model_class: type) -> None:
        """Initialize with Django model class"""
        ...

    @abstractmethod
    def _map_to_domain(self, model_instance: Any) -> TAggregate:
        """Map Django model to domain aggregate"""
        ...

    @abstractmethod
    def _map_to_model(self, aggregate: TAggregate) -> Any:
        """Map domain aggregate to Django model"""
        ...

    async def get_by_id_async(self, id: TId) -> Maybe[TAggregate]:
        """Django ORM implementation of get by ID"""
        ...

    async def add_async(self, aggregate: TAggregate) -> Result[None]:
        """Django ORM implementation of add"""
        ...

    async def update_async(self, aggregate: TAggregate) -> Result[None]:
        """Django ORM implementation of update with optimistic concurrency"""
        ...

# ============================================================================
# EXAMPLE DOMAIN IMPLEMENTATIONS
# ============================================================================

@dataclass(frozen=True)
class OrderId:
    """Example entity ID implementation"""
    value: str

    def __post_init__(self) -> None:
        if not self.value or not self.value.strip():
            raise ValueError("OrderId value cannot be empty")

    def __str__(self) -> str:
        return self.value

    def __eq__(self, other: object) -> bool:
        return isinstance(other, OrderId) and self.value == other.value

    def __hash__(self) -> int:
        return hash(self.value)

@dataclass(frozen=True)
class Money(ValueObject):
    """Example value object implementation"""
    amount: float
    currency: str

    def __post_init__(self) -> None:
        if self.amount < 0:
            raise ValueError("Amount cannot be negative")
        if not self.currency or len(self.currency) != 3:
            raise ValueError("Currency must be 3-character code")

    def get_equality_components(self) -> Iterator[Any]:
        yield self.amount
        yield self.currency.upper()

@dataclass(frozen=True)
class OrderCreated(DomainEventBase):
    """Example domain event implementation"""
    order_id: str
    customer_name: str
    total: Money

class Order(AggregateRoot[OrderId]):
    """Example aggregate root implementation"""

    def __init__(self, id: OrderId, customer_name: str, total: Money) -> None:
        super().__init__(id)
        self._customer_name = customer_name
        self._total = total
        self.add_event(OrderCreated(
            id=str(uuid.uuid4()),
            occurred_at=datetime.utcnow(),
            order_id=str(id),
            customer_name=customer_name,
            total=total
        ))

    @property
    def customer_name(self) -> str:
        return self._customer_name

    @property
    def total(self) -> Money:
        return self._total

class OrderRepository(Repository[Order, OrderId], Protocol):
    """Example repository contract"""

    async def find_by_customer_async(self, customer_name: str) -> List[Order]:
        """Find orders by customer name"""
        ...

# ============================================================================
# MONADIC LAWS CONTRACTS
# ============================================================================

class MonadicLaws:
    """Contract for verifying monadic laws"""

    @staticmethod
    def verify_result_left_identity(value: T, func: Callable[[T], Result[U]]) -> bool:
        """Verify Result left identity law: Result.success(a).bind(f) == f(a)"""
        ...

    @staticmethod
    def verify_result_right_identity(result: Result[T]) -> bool:
        """Verify Result right identity law: m.bind(Result.success) == m"""
        ...

    @staticmethod
    def verify_result_associativity(
        result: Result[T],
        f: Callable[[T], Result[U]],
        g: Callable[[U], Result[Any]]
    ) -> bool:
        """Verify Result associativity law: m.bind(f).bind(g) == m.bind(lambda x: f(x).bind(g))"""
        ...

    @staticmethod
    def verify_maybe_left_identity(value: T, func: Callable[[T], Maybe[U]]) -> bool:
        """Verify Maybe left identity law: Maybe.some(a).bind(f) == f(a)"""
        ...

    @staticmethod
    def verify_maybe_right_identity(maybe: Maybe[T]) -> bool:
        """Verify Maybe right identity law: m.bind(Maybe.some) == m"""
        ...

    @staticmethod
    def verify_maybe_associativity(
        maybe: Maybe[T],
        f: Callable[[T], Maybe[U]],
        g: Callable[[U], Maybe[Any]]
    ) -> bool:
        """Verify Maybe associativity law: m.bind(f).bind(g) == m.bind(lambda x: f(x).bind(g))"""
        ...

# ============================================================================
# PERFORMANCE CONTRACTS
# ============================================================================

class PerformanceRequirements:
    """Contract for performance requirements"""

    MAX_AGGREGATE_OPERATION_TIME_MS = 1.0  # Sub-millisecond operations
    MAX_REPOSITORY_OPERATION_TIME_MS = 200.0  # 200ms p95 for repository operations
    MIN_THROUGHPUT_OPS_PER_SECOND = 1000  # Minimum throughput requirements

    @staticmethod
    def measure_operation_time(operation: Callable[[], T]) -> tuple[T, float]:
        """Measure operation execution time in milliseconds"""
        ...

    @staticmethod
    def verify_memory_efficiency(operation: Callable[[], T], max_allocations: int) -> bool:
        """Verify operation meets memory efficiency requirements"""
        ...

# ============================================================================
# SERIALIZATION CONTRACTS
# ============================================================================

class Serializable(Protocol):
    """Contract for serializable types"""

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to dictionary"""
        ...

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Serializable':
        """Deserialize from dictionary"""
        ...

class JsonSerializable(Serializable, Protocol):
    """Contract for JSON serializable types"""

    def to_json(self) -> str:
        """Serialize to JSON string"""
        ...

    @classmethod
    def from_json(cls, json_str: str) -> 'JsonSerializable':
        """Deserialize from JSON string"""
        ...

# ============================================================================
# TEST CONTRACTS
# ============================================================================

class TestDataBuilder(Generic[T], ABC):
    """Contract for test data builders"""

    @abstractmethod
    def build(self) -> T:
        """Build test data instance"""
        ...

    @abstractmethod
    def with_defaults(self) -> 'TestDataBuilder[T]':
        """Create builder with default values"""
        ...

class RepositoryTestContract(Generic[TAggregate, TId]):
    """Contract for testing repository implementations"""

    async def test_get_by_id_returns_aggregate_when_exists(self) -> None:
        """Test that get_by_id returns aggregate when it exists"""
        ...

    async def test_get_by_id_returns_none_when_not_exists(self) -> None:
        """Test that get_by_id returns None when aggregate doesn't exist"""
        ...

    async def test_add_saves_new_aggregate(self) -> None:
        """Test that add saves new aggregate successfully"""
        ...

    async def test_update_modifies_existing_aggregate(self) -> None:
        """Test that update modifies existing aggregate"""
        ...

    async def test_update_handles_concurrency_conflict(self) -> None:
        """Test that update handles optimistic concurrency conflicts"""
        ...

    async def test_delete_removes_aggregate(self) -> None:
        """Test that delete removes aggregate successfully"""
        ...

    async def test_exists_returns_true_when_exists(self) -> None:
        """Test that exists returns true when aggregate exists"""
        ...

    async def test_exists_returns_false_when_not_exists(self) -> None:
        """Test that exists returns false when aggregate doesn't exist"""
        ...

"""
CONTRACT IMPLEMENTATION REQUIREMENTS:

1. All types must be properly annotated with typing information
2. Generic types must support proper variance
3. All async operations must support cancellation via asyncio
4. All monadic operations must satisfy mathematical laws
5. All repository operations must handle errors gracefully
6. All value objects must be immutable after creation
7. All domain events must be immutable after creation
8. All aggregate operations must maintain consistency
9. All performance requirements must be validated with benchmarks
10. All implementations must support comprehensive testing

DJANGO INTEGRATION REQUIREMENTS:

1. Core library must have zero Django dependencies
2. Django integration must be in separate package
3. Repository implementations must support Django ORM async operations
4. Model mixins must support optimistic concurrency
5. Transaction boundaries must be properly managed
6. Database migrations must be Django-compatible

TESTING REQUIREMENTS:

1. All contracts must have corresponding test suites
2. Property-based testing for monadic laws
3. Performance benchmarking for all operations
4. Integration testing with Django ORM
5. Mock-friendly interfaces for unit testing
6. Comprehensive edge case coverage
"""

File: specs/001-architecture-core-ddd/contracts/core-types-contract.java
// Core Types Contract Specification for Java Spring Implementation
// This file defines the public API contracts for Architecture.Core types
// Package: com.architecture.core

package com.architecture.core.contracts;

import java.io.Serializable;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;
import java.util.function.Supplier;

// =============================================================================
// ENTITY ID ABSTRACTION
// =============================================================================

/**
 * Base interface for all entity identifiers with type safety and comparability.
 * Requirements: Type-safe entity identification, validation, serialization
 */
public interface EntityId<T extends EntityId<T>> extends Comparable<T>, Serializable {
    /**
     * Gets the string representation of the identifier value.
     * MUST NOT return null.
     */
    String getValue();

    /**
     * Validates the identifier format and constraints.
     * MUST return Result.success() for valid identifiers.
     */
    Result<Void> validate();

    @Override
    int compareTo(T other);

    @Override
    boolean equals(Object obj);

    @Override
    int hashCode();
}

// =============================================================================
// DDD ABSTRACTIONS
// =============================================================================

/**
 * Entity with identity-based equality
 * Requirements: FR-003 - Identity-based equality
 */
public abstract class Entity<TId extends EntityId<TId>> {
    private final TId id;

    protected Entity(TId id) {
        this.id = Objects.requireNonNull(id, "Entity ID cannot be null");
    }

    public final TId getId() { return id; }

    @Override
    public final boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Entity<?> entity = (Entity<?>) obj;
        return Objects.equals(id, entity.id);
    }

    @Override
    public final int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return String.format("%s{id=%s}", getClass().getSimpleName(), id);
    }
}

/**
 * Aggregate root with version control and domain events
 * Requirements: FR-001, FR-002 - Version control and event collection
 */
public abstract class AggregateRoot<TId extends EntityId<TId>> extends Entity<TId> {
    private long version;
    private final List<DomainEvent> domainEvents;
    private final Object eventLock = new Object();

    protected AggregateRoot(TId id) {
        super(id);
        this.version = 0;
        this.domainEvents = new ArrayList<>();
    }

    protected AggregateRoot(TId id, long version) {
        super(id);
        this.version = version;
        this.domainEvents = new ArrayList<>();
    }

    public final long getVersion() { return version; }

    public final List<DomainEvent> getDomainEvents() {
        synchronized (eventLock) {
            return Collections.unmodifiableList(new ArrayList<>(domainEvents));
        }
    }

    protected final void addDomainEvent(DomainEvent event) {
        Objects.requireNonNull(event, "Domain event cannot be null");
        synchronized (eventLock) {
            domainEvents.add(event);
        }
    }

    public final void clearDomainEvents() {
        synchronized (eventLock) {
            domainEvents.clear();
        }
    }

    public final void incrementVersion() {
        this.version++;
    }
}

/**
 * Base class for value objects with structural equality
 * Requirements: FR-004, FR-012 - Structural equality and immutability
 */
public abstract class ValueObject {
    /**
     * Returns the components that define equality for this value object.
     * MUST return all fields that participate in equality comparison.
     */
    protected abstract Iterable<Object> getEqualityComponents();

    @Override
    public final boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        ValueObject other = (ValueObject) obj;
        return Objects.equals(getEqualityComponentsList(), other.getEqualityComponentsList());
    }

    @Override
    public final int hashCode() {
        return Objects.hash(getEqualityComponentsList().toArray());
    }

    @Override
    public String toString() {
        return String.format("%s{%s}",
            getClass().getSimpleName(),
            getEqualityComponentsList().stream()
                .map(String::valueOf)
                .collect(Collectors.joining(", "))
        );
    }

    private List<Object> getEqualityComponentsList() {
        List<Object> components = new ArrayList<>();
        getEqualityComponents().forEach(components::add);
        return components;
    }
}

/**
 * Domain event interface with correlation metadata
 * Requirements: FR-005 - Domain event with metadata
 */
public interface DomainEvent {
    UUID getEventId();
    Instant getOccurredAt();
    Optional<String> getCorrelationId();
    Optional<String> getCausationId();
    int getEventVersion();
}

/**
 * Base implementation for domain events
 * Requirements: FR-005 - Standard event metadata
 */
public abstract class DomainEventBase implements DomainEvent {
    private final UUID eventId;
    private final Instant occurredAt;
    private final String correlationId;
    private final String causationId;
    private final int eventVersion;

    protected DomainEventBase() {
        this(null, null, 1);
    }

    protected DomainEventBase(String correlationId, String causationId) {
        this(correlationId, causationId, 1);
    }

    protected DomainEventBase(String correlationId, String causationId, int eventVersion) {
        this.eventId = UUID.randomUUID();
        this.occurredAt = Instant.now();
        this.correlationId = correlationId;
        this.causationId = causationId;
        this.eventVersion = eventVersion;
    }

    @Override
    public final UUID getEventId() { return eventId; }

    @Override
    public final Instant getOccurredAt() { return occurredAt; }

    @Override
    public final Optional<String> getCorrelationId() {
        return Optional.ofNullable(correlationId);
    }

    @Override
    public final Optional<String> getCausationId() {
        return Optional.ofNullable(causationId);
    }

    @Override
    public final int getEventVersion() { return eventVersion; }
}

/**
 * Repository interface for aggregate persistence
 * Requirements: FR-006, FR-007 - Repository with async operations and cancellation
 */
public interface Repository<TAggregate extends AggregateRoot<TId>, TId extends EntityId<TId>> {
    CompletableFuture<Maybe<TAggregate>> getByIdAsync(TId id, CancellationToken cancellationToken);
    CompletableFuture<Result<Void>> addAsync(TAggregate aggregate, CancellationToken cancellationToken);
    CompletableFuture<Result<Void>> updateAsync(TAggregate aggregate, CancellationToken cancellationToken);
    CompletableFuture<Result<Void>> deleteAsync(TId id, CancellationToken cancellationToken);
    CompletableFuture<Boolean> existsAsync(TId id, CancellationToken cancellationToken);
}

// =============================================================================
// FUNCTIONAL TYPES
// =============================================================================

/**
 * Error categories for structured error handling
 * Requirements: FR-010, FR-014 - Error categorization
 */
public enum ErrorCategory {
    DOMAIN("Domain"),
    VALIDATION("Validation"),
    INFRASTRUCTURE("Infrastructure"),
    CONCURRENCY("Concurrency"),
    SECURITY("Security");

    private final String displayName;

    ErrorCategory(String displayName) {
        this.displayName = displayName;
    }

    public String getDisplayName() { return displayName; }

    @Override
    public String toString() { return displayName; }
}

/**
 * Structured error with categorization and metadata
 * Requirements: FR-010 - Error type with categorization
 */
public final class Error {
    private final String code;
    private final String message;
    private final ErrorCategory category;
    private final Map<String, Object> metadata;
    private final Throwable cause;

    public Error(String code, String message, ErrorCategory category,
                 Map<String, Object> metadata, Throwable cause) {
        this.code = Objects.requireNonNull(code, "Error code cannot be null");
        this.message = Objects.requireNonNull(message, "Error message cannot be null");
        this.category = Objects.requireNonNull(category, "Error category cannot be null");
        this.metadata = Map.copyOf(metadata != null ? metadata : Map.of());
        this.cause = cause;
    }

    // Factory methods for different error categories
    public static Error domain(String code, String message) {
        return new Error(code, message, ErrorCategory.DOMAIN, Map.of(), null);
    }

    public static Error validation(String code, String message, Map<String, Object> metadata) {
        return new Error(code, message, ErrorCategory.VALIDATION, metadata, null);
    }

    public static Error infrastructure(String code, String message, Throwable cause) {
        return new Error(code, message, ErrorCategory.INFRASTRUCTURE, Map.of(), cause);
    }

    public static Error concurrency(String code, String message) {
        return new Error(code, message, ErrorCategory.CONCURRENCY, Map.of(), null);
    }

    public static Error security(String code, String message) {
        return new Error(code, message, ErrorCategory.SECURITY, Map.of(), null);
    }

    // Accessors
    public String getCode() { return code; }
    public String getMessage() { return message; }
    public ErrorCategory getCategory() { return category; }
    public Map<String, Object> getMetadata() { return metadata; }
    public Optional<Throwable> getCause() { return Optional.ofNullable(cause); }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Error error = (Error) obj;
        return Objects.equals(code, error.code) &&
               Objects.equals(message, error.message) &&
               category == error.category &&
               Objects.equals(metadata, error.metadata) &&
               Objects.equals(cause, error.cause);
    }

    @Override
    public int hashCode() {
        return Objects.hash(code, message, category, metadata, cause);
    }

    @Override
    public String toString() {
        return String.format("Error{code='%s', message='%s', category=%s}",
            code, message, category);
    }
}

/**
 * Result type for operations that can succeed or fail
 * Requirements: FR-008, FR-011, FR-013 - Result with monadic operations
 */
public final class Result<T> {
    private final T value;
    private final Error error;
    private final boolean isSuccess;

    private Result(T value, Error error, boolean isSuccess) {
        this.value = value;
        this.error = error;
        this.isSuccess = isSuccess;
    }

    // Factory methods
    public static <T> Result<T> success(T value) {
        Objects.requireNonNull(value, "Success value cannot be null");
        return new Result<>(value, null, true);
    }

    public static <T> Result<T> failure(Error error) {
        Objects.requireNonNull(error, "Error cannot be null");
        return new Result<>(null, error, false);
    }

    public static <T> Result<T> failure(String code, String message) {
        return failure(Error.domain(code, message));
    }

    // State queries
    public boolean isSuccess() { return isSuccess; }
    public boolean isFailure() { return !isSuccess; }

    public T getValue() {
        if (!isSuccess) {
            throw new IllegalStateException("Cannot get value from failed result");
        }
        return value;
    }

    public Error getError() {
        if (isSuccess) {
            throw new IllegalStateException("Cannot get error from successful result");
        }
        return error;
    }

    // Monadic operations (Requirements: FR-011 - Monadic laws)
    public <U> Result<U> map(Function<? super T, ? extends U> mapper) {
        Objects.requireNonNull(mapper, "Mapper function cannot be null");
        if (isSuccess()) {
            try {
                return Result.success(mapper.apply(getValue()));
            } catch (Exception e) {
                return Result.failure(Error.infrastructure("Mapping.Failed", e.getMessage(), e));
            }
        }
        return Result.failure(getError());
    }

    public <U> Result<U> bind(Function<? super T, Result<U>> binder) {
        Objects.requireNonNull(binder, "Binder function cannot be null");
        if (isSuccess()) {
            try {
                return binder.apply(getValue());
            } catch (Exception e) {
                return Result.failure(Error.infrastructure("Binding.Failed", e.getMessage(), e));
            }
        }
        return Result.failure(getError());
    }

    public <U> U match(Function<? super T, ? extends U> onSuccess, Function<Error, ? extends U> onFailure) {
        Objects.requireNonNull(onSuccess, "Success handler cannot be null");
        Objects.requireNonNull(onFailure, "Failure handler cannot be null");
        return isSuccess() ? onSuccess.apply(getValue()) : onFailure.apply(getError());
    }

    // Utility methods
    public T getValueOrDefault(T defaultValue) {
        return isSuccess() ? getValue() : defaultValue;
    }

    public T getValueOrThrow() {
        if (isSuccess()) {
            return getValue();
        }
        throw new ResultException(getError());
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Result<?> result = (Result<?>) obj;
        return isSuccess == result.isSuccess &&
               Objects.equals(value, result.value) &&
               Objects.equals(error, result.error);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value, error, isSuccess);
    }

    @Override
    public String toString() {
        return isSuccess()
            ? String.format("Success{value=%s}", value)
            : String.format("Failure{error=%s}", error);
    }
}

/**
 * Maybe type for optional values
 * Requirements: FR-009, FR-011 - Optional values with monadic operations
 */
public final class Maybe<T> {
    private static final Maybe<?> NONE = new Maybe<>(null, false);

    private final T value;
    private final boolean hasValue;

    private Maybe(T value, boolean hasValue) {
        this.value = value;
        this.hasValue = hasValue;
    }

    // Factory methods
    public static <T> Maybe<T> some(T value) {
        Objects.requireNonNull(value, "Value cannot be null");
        return new Maybe<>(value, true);
    }

    @SuppressWarnings("unchecked")
    public static <T> Maybe<T> none() {
        return (Maybe<T>) NONE;
    }

    public static <T> Maybe<T> fromNullable(T value) {
        return value != null ? some(value) : none();
    }

    // State queries
    public boolean hasValue() { return hasValue; }
    public boolean isEmpty() { return !hasValue; }

    public T getValue() {
        if (!hasValue) {
            throw new IllegalStateException("Maybe has no value");
        }
        return value;
    }

    // Monadic operations (Requirements: FR-011 - Monadic laws)
    public <U> Maybe<U> map(Function<? super T, ? extends U> mapper) {
        Objects.requireNonNull(mapper, "Mapper function cannot be null");
        if (hasValue()) {
            try {
                U result = mapper.apply(getValue());
                return result != null ? Maybe.some(result) : Maybe.none();
            } catch (Exception e) {
                return Maybe.none();
            }
        }
        return Maybe.none();
    }

    public <U> Maybe<U> bind(Function<? super T, Maybe<U>> binder) {
        Objects.requireNonNull(binder, "Binder function cannot be null");
        if (hasValue()) {
            try {
                return binder.apply(getValue());
            } catch (Exception e) {
                return Maybe.none();
            }
        }
        return Maybe.none();
    }

    // Value extraction
    public T orElse(T defaultValue) {
        return hasValue() ? getValue() : defaultValue;
    }

    public T orElseGet(Supplier<? extends T> defaultSupplier) {
        Objects.requireNonNull(defaultSupplier, "Default supplier cannot be null");
        return hasValue() ? getValue() : defaultSupplier.get();
    }

    public <X extends Throwable> T orElseThrow(Supplier<? extends X> exceptionSupplier) throws X {
        Objects.requireNonNull(exceptionSupplier, "Exception supplier cannot be null");
        if (hasValue()) {
            return getValue();
        }
        throw exceptionSupplier.get();
    }

    // Conversion methods (Requirements: FR-013 - Conversions)
    public Optional<T> toOptional() {
        return hasValue() ? Optional.of(getValue()) : Optional.empty();
    }

    public Result<T> toResult(Error errorIfEmpty) {
        return hasValue() ? Result.success(getValue()) : Result.failure(errorIfEmpty);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Maybe<?> maybe = (Maybe<?>) obj;
        return hasValue == maybe.hasValue && Objects.equals(value, maybe.value);
    }

    @Override
    public int hashCode() {
        return hasValue ? Objects.hash(value) : 0;
    }

    @Override
    public String toString() {
        return hasValue ? String.format("Some{value=%s}", value) : "None";
    }
}

// =============================================================================
// SUPPORT TYPES
// =============================================================================

/**
 * Exception that wraps a structured Error for integration with exception-based APIs
 */
public class ResultException extends RuntimeException {
    private final Error error;

    public ResultException(Error error) {
        super(error.getMessage(), error.getCause().orElse(null));
        this.error = Objects.requireNonNull(error, "Error cannot be null");
    }

    public Error getError() {
        return error;
    }

    @Override
    public String toString() {
        return String.format("ResultException{error=%s}", error);
    }
}

/**
 * Token for cancelling asynchronous operations
 */
public interface CancellationToken {
    boolean isCancellationRequested();
    void throwIfCancellationRequested() throws OperationCancelledException;

    static CancellationToken none() {
        return NonCancellationToken.INSTANCE;
    }
}

/**
 * Exception thrown when an operation is cancelled
 */
public class OperationCancelledException extends RuntimeException {
    public OperationCancelledException() {
        super("Operation was cancelled");
    }

    public OperationCancelledException(String message) {
        super(message);
    }
}

/**
 * Non-cancelling token implementation
 */
final class NonCancellationToken implements CancellationToken {
    static final CancellationToken INSTANCE = new NonCancellationToken();

    private NonCancellationToken() {}

    @Override
    public boolean isCancellationRequested() {
        return false;
    }

    @Override
    public void throwIfCancellationRequested() {
        // No-op
    }
}

// =============================================================================
// CONTRACT VERIFICATION INTERFACES
// =============================================================================

/**
 * Contract test interfaces that must be implemented to verify API compliance
 * Requirements: All functional requirements (FR-001 through FR-017)
 */
public interface ContractTests {
    // Entity ID contracts
    void testEntityIdValidation();
    void testEntityIdComparison();
    void testEntityIdSerialization();

    // Entity contracts (FR-003)
    void testEntityIdentityEquality();
    void testEntityHashCodeStability();

    // Value Object contracts (FR-004, FR-012)
    void testValueObjectStructuralEquality();
    void testValueObjectHashCodeStability();
    void testValueObjectImmutability();
    void testValueObjectNullHandling();

    // Aggregate Root contracts (FR-001, FR-002)
    void testAggregateVersionIncrement();
    void testAggregateEventCollection();
    void testAggregateEventClearing();
    void testAggregateThreadSafety();

    // Domain Event contracts (FR-005)
    void testDomainEventMetadata();
    void testDomainEventEquality();

    // Repository contracts (FR-006, FR-007)
    void testRepositoryAsyncOperations();
    void testRepositoryCancellation();
    void testRepositoryResultHandling();

    // Result contracts (FR-008, FR-011, FR-013)
    void testResultMonadicLaws();
    void testResultErrorHandling();
    void testResultTypeConversions();

    // Maybe contracts (FR-009, FR-011)
    void testMaybeMonadicLaws();
    void testMaybeNullHandling();
    void testMaybeConversions();

    // Error contracts (FR-010, FR-014)
    void testErrorCategorization();
    void testErrorMetadata();
    void testErrorSerialization();
}

/**
 * Performance contract requirements
 * Requirements: Performance goals from technical context
 */
public interface PerformanceContracts {
    // Memory efficiency requirements
    void testResultMemoryUsage();
    void testMaybeMemoryUsage();
    void testValueObjectMemoryUsage();

    // Execution time requirements
    void testValueObjectEqualityPerformance();
    void testAggregateEventPerformance();
    void testResultMonadicOperationsPerformance();
}